Add AccessorArg_t View Ctor Property

crtrott opened the pull request:
This allows passing through a size_t from view construction to the accessor. It is used as part of the View customization. We later want to extent this to be templated, so we can pass in other stuff than size_t, but I only need size_t for Sacado.

Example view construction (but you also need the view customization hook thingy to make the special accessor):
```c++
Kokkos::View<SpecialScalar*> a(Kokkos::view_alloc("A", Kokkos::Impl::AccessorArg_t{5}), N);
```

Note this is like the customization hook in impl namespace so we can refine it.

crtrott committed:
Add AccessorArg_t View Ctor Property

This allows passing through a size_t from view construction to
the accessor. It is used as part of the View customization.
We later want to extent this to be templated, so we can pass in
other stuff than size_t, but I only need size_t for Sacado.

Signed-off-by: Christian Trott <crtrott@sandia.gov>

masterleinad commented on core/src/View/Kokkos_BasicView.hpp:
```diff
@@ -385,23 +385,67 @@ class BasicView {
+                       typename mdspan_type::mapping_type> const &arg_mapping)
+      : BasicView(arg_prop, arg_mapping, accessor_type{}) {}
+
+  template <class... P>
+
+  explicit inline BasicView(
```
```suggestion
  template <class... P>
  explicit inline BasicView(
```

crtrott replied:
Done

masterleinad reviewed (approved):
Looks reasonable assuming that passing through a single `size_t` really sufficient to make things work with `Sacado`. Again, I would rather have `AccessorArg_t` in a public namespace but it makes sense to keep that uniform with `ViewCustomArguments`.

masterleinad commented:
```
/var/jenkins/workspace/Kokkos_PR-8112/install/include/View/Kokkos_ViewCtor.hpp(41): error: __device__ annotation is ignored on a function("AccessorArg_t") that is explicitly defaulted on its first declaration

/var/jenkins/workspace/Kokkos_PR-8112/install/include/View/Kokkos_ViewCtor.hpp(41): error: __host__ annotation is ignored on a function("AccessorArg_t") that is explicitly defaulted on its first declaration
```

masterleinad commented on core/src/View/Kokkos_ViewCtor.hpp:
```diff
@@ -27,6 +27,22 @@ struct SequentialHostInit_t {};
+  size_t val{};
+
+ public:
+  KOKKOS_FUNCTION AccessorArg_t() = default;
```
```suggestion
  KOKKOS_DEFAULTED_FUNCTION AccessorArg_t() = default;
```

JBludau replied:
Would it be ok to convert this to an aggregate?

crtrott replied:
I started that way and then I got annoyed that I can't write `AccessorArg_t(0)` lol

crtrott replied:
anyway I did change it to an aggregate again.

dalg24 commented on core/unit_test/view/TestViewCustomizationAccessorArg.hpp:
```diff
@@ -0,0 +1,125 @@
+      Kokkos::Impl::ReferenceCountedDataHandle<ElementType, MemorySpace>;
+  using offset_policy = TestAccessor;
+
+  // View expects this from accessors right now
```
If this intended as something to re-evaluate later, then consider using a TODO_FOO or FIXME_BAR markup to increase the chances we actually follow through with follow up work.

crtrott replied:
It is not intended to change for now. It's a bigger discussion we have to have.

dalg24 commented on core/unit_test/view/TestViewCustomizationAccessorArg.hpp:
```diff
@@ -0,0 +1,125 @@
+#ifndef KOKKOS_ENABLE_OPENACC
+      const data_handle_type& p,
+#else
+      // FIXME OpenACC: illegal address when passing by reference
```
```suggestion
      // FIXME_OPENACC illegal address when passing by reference
```

crtrott replied:
Done

dalg24 commented on core/unit_test/view/TestViewCustomizationAccessorArg.hpp:
```diff
@@ -0,0 +1,125 @@
+#ifndef KOKKOS_ENABLE_OPENACC
+      const data_handle_type& p,
+#else
+      // FIXME OpenACC: illegal address when passing by reference
```
```suggestion
      // FIXME_OPENACC illegal address when passing by reference
```

crtrott replied:
Done

dalg24 commented on core/unit_test/view/TestViewCustomizationAccessorArg.hpp:
```diff
@@ -0,0 +1,125 @@
+      data_handle_type p,
+#endif
+      size_t i) const noexcept {
+    return *(p + (i % value));
```
So essentially we are mocking up this recurring use case where the element type is an array and flattening (e.g. Sacado or Stokhos).
Can we come up with better names for "TestAccessor" and "value" or provide a short preamble giving a little context on what this test is about?  I am sure the future reader will appreciate.

JBludau replied:
how about being explicit that this is linked to array flattening in sacado?

crtrott replied:
It is not. I just did something with value. This doesn't do array flattening or anything else really related to Sacado.

dalg24 commented on core/unit_test/view/TestViewCustomizationAccessorArg.hpp:
```diff
@@ -0,0 +1,125 @@
+  ASSERT_EQ(num_error, 0);
+}
+
+TEST(TEST_CATEGORY, view_customization_accessor_arg) { test_accessor_arg(); }
```
Why not inlining the test here?

crtrott replied:
because GTEST / CUDA LAMBDA doesn't play together.

dalg24 commented on core/unit_test/view/TestViewCustomizationAccessorArg.hpp:
```diff
@@ -0,0 +1,125 @@
+          Foo::TestAccessor<Foo::Bar, typename TEST_EXECSPACE::memory_space>>);
+
+  view_t a("A", 10);
+  ASSERT_EQ(a.accessor().value, 0ul);
```
maybe comment that this is a "cheap" way of comparing accessors.

dalg24 commented on core/unit_test/CMakeLists.txt:
```diff
@@ -367,6 +367,16 @@ foreach(Tag Threads;Serial;OpenMP;Cuda;HPX;OpenMPTarget;OpenACC;HIP;SYCL)
       endforeach()
+      if(NOT Kokkos_ENABLE_IMPL_VIEW_LEGACY)
+        foreach(Name ViewCustomizationAccessorArg)
+          set(file ${dir}/Test${Tag}_View_${Name}.cpp)
+          # Write to a temporary intermediate file and call configure_file to avoid
+          # updating timestamps triggering unnecessary rebuilds on subsequent cmake runs.
+          file(WRITE ${dir}/dummy.cpp "#include <Test${Tag}_Category.hpp>\n" "#include <view/Test${Name}.hpp>\n")
+          configure_file(${dir}/dummy.cpp ${file})
+          list(APPEND ${Tag}_VIEWSUPPORT ${file})
```
Instead of duplicating that block, why are we not using a variable for the list of `Name`s to iterate over?

crtrott replied:
done

dalg24 commented on core/src/View/Kokkos_ViewCtor.hpp:
```diff
@@ -27,6 +27,22 @@ struct SequentialHostInit_t {};
+// But that adds some more complication, preventing simply
+// copying the code from the above properties.
+// For Sacado we only need a size_t for fad_size
+// We can template this later in a separate change.
```
Can you briefly explain why we want an argument to the accessor constructor rather than store the accessor directly?

crtrott replied:
Because Sacado (or other) users wouldn't even know what the accessor is. The accessor can be an implementation detail of Sacado. You just need to know that if you use a FAD type you have to pass `AccessorArg{fad_size}` to `view_alloc`. Also the accessor needs even more information. In fact I will next introduce the lovely invention of `construct_accessor_from_accessor_arg_and_mapping` ...

dalg24 commented on core/src/View/Kokkos_BasicView.hpp:
```diff
@@ -385,23 +385,67 @@ class BasicView {
  public:
+  // check for accessor_arg for managed view ctors
+  template <class... P>
+  explicit inline BasicView(
```
The `inline` specifier here is pointless (I see we have it at other places so do whatever you want)

crtrott replied:
Removed

dalg24 commented on core/src/View/Kokkos_BasicView.hpp:
```diff
@@ -385,23 +385,67 @@ class BasicView {
+                       typename mdspan_type::mapping_type> const &arg_mapping,
+      const accessor_type &arg_accessor)
+      : BasicView(create_data_handle(arg_prop, arg_mapping), arg_mapping,
+                  arg_accessor) {}
```
Do we need that constructor that takes directly the accessor as trailing argument or can we live with the ViewCtorProp that holds an accessor?

What if arg_prop has an accessor arg?  Why is it ok to just ignore it?
(Same question below)

nmm0 replied:
This constructor should have a constraint that arg_prop doesn't have an accessor arg

crtrott replied:
ViewCtorProp doesn't currently hold accessors - and it also doesn't make sense. I mean in my mind it's the same as saying we pass mapping through CtorProp.  Also BasicView at its most fundamental is created from data_handle, mapping, and accessor - its constructors should express that.

crtrott replied:
If we constrain it not containing ArgAccessor I need to remove it in the above ctor from CtorProp.

crtrott replied:
Or we add another 4 or 8 ctors to deal with the combinatorical situation

crtrott replied:
By duplicating the data_handle creation logic, and removing again the ability to call constructors such as `View<..>(view_alloc("A"), mapping, accessor)` (which we do not have in develop) I was able to disallow the simultaneous existence of CtorProp with AccessorArg and an accessor, other than in a private ctor.

JBludau replied:
Maybe I don't see it but I still don't understand why this ctor does *not* need `!Impl::ViewCtorProp<P...>::has_accessor_arg`

crtrott replied:
its a private one which only is getting called from the the one which had an accessor_arg

nmm0 commented on core/src/View/Kokkos_BasicView.hpp:
```diff
@@ -385,23 +385,67 @@ class BasicView {
  public:
+  // check for accessor_arg for managed view ctors
+  template <class... P>
+  explicit inline BasicView(
```
```suggestion
  explicit BasicView(
```

nmm0 replied:
Not familiar enough with Sacado -- does this need `KOKKOS_FUNCTION`

crtrott replied:
It can't: this is the allocation CTor (CtorProp does not contain a pointer).

nmm0 commented on core/src/View/Kokkos_BasicView.hpp:
```diff
@@ -385,23 +385,67 @@ class BasicView {
+                       typename mdspan_type::mapping_type> const &arg_mapping)
+      : BasicView(arg_prop, arg_mapping, accessor_type{}) {}
+
+  template <class... P>
+
+  explicit inline BasicView(
```
```suggestion
  template <class... P>
  explicit KOKKOS_FUNCTION BasicView(
```

This is the non-arg constructor right -- so it needs `KOKKOS_FUNCTION`?

crtrott replied:
No: it's a ctor which allocates so it can't have KOKKOS_FUNCTION (!has_pointer).

nmm0 commented on core/src/View/Kokkos_BasicView.hpp:
```diff
@@ -385,23 +385,67 @@ class BasicView {
       std::enable_if_t<Impl::ViewCtorProp<P...>::has_pointer,
-                       typename mdspan_type::mapping_type> const &arg_mapping)
+                       typename mdspan_type::mapping_type> const &arg_mapping,
+      const accessor_type &arg_accessor)
```
Same here

crtrott committed:
Address Review comments

Signed-off-by: Christian Trott <crtrott@sandia.gov>

crtrott committed:
Redo cmake logic for ViewSupport Test

Signed-off-by: Christian Trott <crtrott@sandia.gov>

crtrott committed:
Work around MSVC+CUDA issue with multiple bases and pack expansion

Signed-off-by: Christian Trott <crtrott@sandia.gov>

JBludau commented:
Retest this please

crtrott committed:
Disable View Customization Test for OpenACC

Fails with a similar issue as the one we had where passing data
handles by reference would result in illegal memory accesses.
I did the same here, but it still doesn't work.

Let this for someone else to figure out.

crtrott commented:
OpenACC failed in previous test round

crtrott commented:
The two failures are unrelated: SNL_HIP there seems to be a machine issue with one of the nodes, the other one (independent instances) seems a race condition.
